### PR TITLE
extend resource state

### PR DESCRIPTION
**Pull request by fgimenez**
Towards https://github.com/giantswarm/giantswarm/issues/2023

Split from #501

### CODE REVIEW

File: service/resource/cloudformation/main_stack.go [hunk 1]
```diff
@@ -2,19 +2,27 @@ package cloudformation
...
 	"github.com/giantswarm/aws-operator/service/key"
 	"github.com/giantswarm/awstpr"
 )
 
 func newMainStack(customObject awstpr.CustomObject) (StackState, error) {
 	stackName := key.MainStackName(customObject)
+	workers := len(customObject.Spec.AWS.Workers)
+	var imageID string
+	if workers > 0 {
+		imageID = customObject.Spec.AWS.Workers[0].ImageID
```
Comment (rossf7) [inline service/resource/cloudformation/main_stack.go]: Let's create an issue for restructuring awstpr and have a comment here.

Comment (fgimenez) [reply service/resource/cloudformation/main_stack.go]: Sure, done

File: service/resource/cloudformation/resource.go [hunk 2]
```diff
@@ -110,3 +110,13 @@ func toDeleteStackInput(v interface{}) (awscloudformation.DeleteStackInput, erro
...
 }
+
+func getStackOutputValue(outputs []*awscloudformation.Output, key string) (string, error) {
+	for _, o := range outputs {
+		if *o.OutputKey == key {
+			return *o.OutputValue, nil
+		}
+	}
+
+	return "", notFoundError
```
Comment (xh3b4sd) [inline service/resource/cloudformation/resource.go]: Please always mask errors.

Comment (fgimenez) [reply service/resource/cloudformation/resource.go]: Sure, i'm doing it after receiving the error https://github.com/giantswarm/aws-operator/pull/502/files#diff-4313bcb330a12c7c826a88a7090a0905R46 what is the criteria here, always mask at the lower level and let it bubble up untouched? or remask at each stage?

Comment (xh3b4sd) [reply service/resource/cloudformation/resource.go]: Masking everywhere.

Comment (fgimenez) [reply service/resource/cloudformation/resource.go]: Ok done, could you please elaborate on the reason for masking more than once?

Comment (xh3b4sd) [reply service/resource/cloudformation/resource.go]: Imagine your code handles errors in three places. When you do not mask you get an error message. How do you find the problem and the code that produced the error? Similar problem when masking only once. When you mask at the end you have the same problem than not masking at all. When masking somewhere in the middle of the stack you know where the error was within one place of the call stack but you do not know where it originates from nor where it was handled finally. When you mask once where the error can be produced you also miss the rest of the call stack which is often required to debug and solve an issue or at least to understand the situation.

Comment (fgimenez) [reply service/resource/cloudformation/resource.go]: Ok, I agree with the reasons for not masking in places different where the error originates, but masking it at creation the full stack of how you got there should be available AIUI, am i missing something?

Anyway this question is not related to this PR, should we go ahead with it?

Comment (xh3b4sd) [reply service/resource/cloudformation/resource.go]: I am happy to have better ways of doing it. Feel free to provide alternatives. Until then we should keep the convention up as it is.

**Pull request merged**